pkgs/build-support/fetchnodemodules/default.nix

```diff
@@ -0,0 +1,102 @@
+{ lib, stdenvNoCC, makeWrapper, nodejs }:
+
+{ src
+, hash ? ""
+, sha256 ? ""
+, name ? null
+, production ? true
+, runScripts ? false
+, makeTarball ? true
+, preferLocalBuild ? true
+, exposeBin ? false
+, npmFlags ? ""
+, ...
+} @ args:
+
+assert makeTarball -> !exposeBin;
+let
+  suffix = if makeTarball then ".tar.gz" else "";
+  name_ = if name == null then "${src.name}-node_modules${suffix}" else name;
+  nodejs_ = if args ? nodejs then args.nodejs else nodejs;
+
+  hash_ =
+    if hash != "" then { outputHashAlgo = null; outputHash = hash; }
+    else if sha256 != "" then { outputHashAlgo = "sha256"; outputHash = sha256; }
+    else throw "fetchNodeModules requires a hash for ${name_}";
+
+  customArgs = [ "name" "hash" "sha256" ];
+in stdenvNoCC.mkDerivation ({
+  inherit src production runScripts makeTarball preferLocalBuild;
+
+  name = name_;
+  nativeBuildInputs = [ nodejs_ makeWrapper ];
+
+  phases = "unpackPhase patchPhase buildPhase installPhase";
+
+  buildPhase = ''
+    if [[ ! -f package.json ]]; then
+        echo
+        echo "ERROR: The package.json file doesn't exist"
+        echo
+        exit 1
+    fi
+    if [[ ! -f package-lock.json ]]; then
+        echo
+        echo "ERROR: The package-lock.json file doesn't exist"
+        echo
+        echo "package-lock.json is required to make sure the node_modules"
+        echo "set we download never changes."
+        exit 1
+    fi
+
+    export SOURCE_DATE_EPOCH=1
+    export npm_config_cache=/tmp
```

Comment:
isn't it a problem to refer to temp directly ? (just wondering)
lots of other places have use `mktemp`

Reply:
In the darwin sandbox you can't use `/tmp` iirc. `mktemp` uses the right tmp from the environment variables and creates a unique directory, which may prevent a surprise one day.